halo2/src/circuit_builder.rs

```diff
@@ -152,9 +155,21 @@ impl<'a, T: FieldElement, F: PrimeField<Repr = [u8; 32]>> Circuit<F> for PowdrCi
         let advice = analyzed
             .committed_polys_in_source_order()
             .iter()
-            .flat_map(|(symbol, _)| symbol.array_elements())
-            // TODO(challenges): Set correct phase
-            .map(|(name, _)| (name.clone(), meta.advice_column_in(FirstPhase)))
+            .flat_map(|(symbol, _)| {
+                symbol
+                    .array_elements()
+                    .map(|(name, _)| (name, symbol.stage))
+            })
+            .map(|(name, stage)| {
+                let stage = stage.unwrap_or(0);
+                let col = match stage {
+                    0 => meta.advice_column_in(FirstPhase),
+                    1 => meta.advice_column_in(SecondPhase),
+                    2 => meta.advice_column_in(ThirdPhase),
+                    _ => panic!("Unknown phase: {}", stage),
```

Comment:
Shouldn't this rather be "stage too large for halo2 backend"? And could you move the match inside the `advince_column_in` call?